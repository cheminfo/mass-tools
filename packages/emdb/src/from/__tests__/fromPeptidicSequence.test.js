'use strict';

const DBManager = require('../..');

describe('fromPeptidicSequence', () => {
  it('AAKK', async () => {
    let dbManager = new DBManager();
    await dbManager.fromPeptidicSequence('AAKK', {
      allowNeutralLoss: false,
      protonation: false,
      protonationPH: 7,
      ionizations: 'H+,Na+',
      fragmentation: {
        a: true,
      },
      filter: {
        minMSEM: 100,
        maxMSEM: 300,
        targetMass: 150, // just to test, this is useless with precision 1e6
        precision: 1e6,
      },
    });

    const peptidic = dbManager.databases.peptidic.sort(
      (a, b) => a.ms.em - b.ms.em,
    );

    expect(peptidic).toHaveLength(2);
    expect(peptidic).toMatchSnapshot();
  });

  it('AAKK with callback', async () => {
    let dbManager = new DBManager();
    await dbManager.fromPeptidicSequence('AAKK', {
      allowNeutralLoss: false,
      protonation: false,
      protonationPH: 7,
      ionizations: 'H+,Na+',
      fragmentation: {
        a: true,
      },
      filter: {
        callback: (value) => {
          return value.parts[0].includes('AlaAlaLys');
        },
      },
    });

    const peptidic = dbManager.databases.peptidic.sort(
      (a, b) => a.ms.em - b.ms.em,
    );

    expect(peptidic).toHaveLength(4);
    expect(peptidic).toMatchSnapshot();
  });

  it('AAKKKKKKKKKKKKKKKKKK allowNeutralLoss', async () => {
    let dbManager = new DBManager();
    await dbManager.fromPeptidicSequence('AAKKKKKKKKKKKKKKKKKK', {
      allowNeutralLoss: true,
      protonation: false,
      protonationPH: 7,
      limit: 1000000,
      ionizations: 'H+,Na+',
      fragmentation: {
        a: true,
      },
      filter: {
        minMSEM: 100,
        maxMSEM: 300,
        targetMass: 150, // just to test, this is useless with precision 1e6
        precision: 1e6,
      },
    });

    const peptidic = dbManager.databases.peptidic.sort(
      (a, b) => a.ms.em - b.ms.em,
    );

    expect(peptidic).toHaveLength(23);
    expect(peptidic).toMatchSnapshot();
  });

  it('AAKKKKKKKKKKKKKKKKKK filter callback', async () => {
    let dbManager = new DBManager();
    await dbManager.fromPeptidicSequence('AAKKKKKKKKKKKKKKKKKK', {
      ionizations: 'H+,Na+',
      fragmentation: {
        a: true,
      },
      filter: {
        callback: (entry) => entry.atoms.C * 2 === entry.atoms.H,
      },
    });

    expect(dbManager.databases.peptidic).toHaveLength(2);
  });

  it('AAKKKKKK allowNeutralLoss limit: 1000', async () => {
    let dbManager = new DBManager();
    await expect(
      dbManager.fromPeptidicSequence('AAKKKKKKKKK', {
        allowNeutralLoss: true,
        protonation: false,
        protonationPH: 7,
        limit: 100,
        ionizations: 'H+,Na+',
        fragmentation: {
          a: true,
        },

        filter: {
          minMSEM: 100,
          maxMSEM: 300,
          targetMass: 150, // just to test, this is useless with precision 1e6
          precision: 1e6,
        },
      }),
    ).rejects.toMatchInlineSnapshot(
      `[Error: processRange generates to many fragments (over 100)]`,
    );
  });

  it('Linked AA(H-1#1)AA,GG(H-1#1)GG', async () => {
    let dbManager = new DBManager();
    await dbManager.fromPeptidicSequence('AA(H-1#1)AA,GG(H-1#2)GG', {
      links: { filter: true },
      mfsArray: ['#1C6H4#2'],
      fragmentation: {
        a: true,
      },
    });

    const peptidic = dbManager.databases.peptidic
      .sort((a, b) => a.ms.em - b.ms.em)
      .map((item) => item.parts);
    expect(peptidic).toHaveLength(9);
    expect(peptidic).toMatchSnapshot();
  });

  it('Optional linked AA(H-1#1)A,GG(H-1#1)G', async () => {
    let dbManager = new DBManager();
    await dbManager.fromPeptidicSequence('AA(H-1#1)A,GG(H-1#2)G', {
      links: { filter: true },
      mfsArray: ['#1C6H4#2,'],
      fragmentation: {
        a: true,
      },
    });

    const peptidic = dbManager.databases.peptidic
      .sort((a, b) => a.ms.em - b.ms.em)
      .map((item) => item.parts);
    expect(peptidic).toHaveLength(6);
    expect(peptidic).toMatchSnapshot();
  });
});
