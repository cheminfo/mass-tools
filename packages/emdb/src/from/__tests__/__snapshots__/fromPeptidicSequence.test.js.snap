// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fromPeptidicSequence AAKK 1`] = `
Array [
  Object {
    "atoms": Object {
      "C": 11,
      "H": 23,
      "N": 4,
      "O": 2,
    },
    "charge": 1,
    "comment": "a3",
    "em": 243.18210099815,
    "ionization": Object {
      "charge": 1,
      "em": 1.00782503223,
      "mf": "H+",
    },
    "mf": "C11H23N4O2(+1)",
    "ms": Object {
      "charge": 2,
      "delta": 27.905585564719075,
      "em": 122.09441443528092,
      "ionization": "H+",
      "ppm": 228557.42986925173,
    },
    "mw": 243.32635490179035,
    "parts": Array [
      "HAlaAlaLysC-1O-1(+1)",
    ],
    "sequence": "AAK",
    "unsaturation": 2,
  },
  Object {
    "atoms": Object {
      "C": 11,
      "H": 23,
      "N": 4,
      "O": 2,
    },
    "charge": 1,
    "comment": "a3",
    "em": 243.18210099815,
    "ionization": Object {
      "charge": 1,
      "em": 22.989769282,
      "mf": "Na+",
    },
    "mf": "C11H23N4O2(+1)",
    "ms": Object {
      "charge": 2,
      "delta": 16.914613439834056,
      "em": 133.08538656016594,
      "ionization": "Na+",
      "ppm": 127095.94852615324,
    },
    "mw": 243.32635490179035,
    "parts": Array [
      "HAlaAlaLysC-1O-1(+1)",
    ],
    "sequence": "AAK",
    "unsaturation": 2,
  },
]
`;

exports[`fromPeptidicSequence AAKK with callback 1`] = `
Array [
  Object {
    "atoms": Object {
      "C": 11,
      "H": 23,
      "N": 4,
      "O": 2,
    },
    "charge": 1,
    "comment": "a3",
    "em": 243.18210099815,
    "ionization": Object {
      "charge": 1,
      "em": 1.00782503223,
      "mf": "H+",
    },
    "mf": "C11H23N4O2(+1)",
    "ms": Object {
      "charge": 2,
      "em": 122.09441443528092,
      "ionization": "H+",
    },
    "mw": 243.32635490179035,
    "parts": Array [
      "HAlaAlaLysC-1O-1(+1)",
    ],
    "sequence": "AAK",
    "unsaturation": 2,
  },
  Object {
    "atoms": Object {
      "C": 11,
      "H": 23,
      "N": 4,
      "O": 2,
    },
    "charge": 1,
    "comment": "a3",
    "em": 243.18210099815,
    "ionization": Object {
      "charge": 1,
      "em": 22.989769282,
      "mf": "Na+",
    },
    "mf": "C11H23N4O2(+1)",
    "ms": Object {
      "charge": 2,
      "em": 133.08538656016594,
      "ionization": "Na+",
    },
    "mw": 243.32635490179035,
    "parts": Array [
      "HAlaAlaLysC-1O-1(+1)",
    ],
    "sequence": "AAK",
    "unsaturation": 2,
  },
  Object {
    "atoms": Object {
      "C": 18,
      "H": 36,
      "N": 6,
      "O": 5,
    },
    "charge": 0,
    "em": 416.27471828471,
    "ionization": Object {
      "charge": 1,
      "em": 1.00782503223,
      "mf": "H+",
    },
    "mf": "C18H36N6O5",
    "ms": Object {
      "charge": 1,
      "em": 417.2819947370309,
      "ionization": "H+",
    },
    "mw": 416.5163571775086,
    "parts": Array [
      "HAlaAlaLysLysOH",
    ],
    "sequence": "AAKK",
    "unsaturation": 4,
  },
  Object {
    "atoms": Object {
      "C": 18,
      "H": 36,
      "N": 6,
      "O": 5,
    },
    "charge": 0,
    "em": 416.27471828471,
    "ionization": Object {
      "charge": 1,
      "em": 22.989769282,
      "mf": "Na+",
    },
    "mf": "C18H36N6O5",
    "ms": Object {
      "charge": 1,
      "em": 439.2639389868009,
      "ionization": "Na+",
    },
    "mw": 416.5163571775086,
    "parts": Array [
      "HAlaAlaLysLysOH",
    ],
    "sequence": "AAKK",
    "unsaturation": 4,
  },
]
`;

exports[`fromPeptidicSequence AAKK with filterFct 1`] = `
Array [
  Object {
    "atoms": Object {
      "C": 11,
      "H": 23,
      "N": 4,
      "O": 2,
    },
    "charge": 1,
    "comment": "a3",
    "em": 243.18210099815,
    "ionization": Object {
      "charge": 1,
      "em": 1.00782503223,
      "mf": "H+",
    },
    "mf": "C11H23N4O2(+1)",
    "ms": Object {
      "charge": 2,
      "delta": 27.905585564719075,
      "em": 122.09441443528092,
      "ionization": "H+",
      "ppm": 228557.42986925173,
    },
    "mw": 243.32635490179035,
    "parts": Array [
      "HAlaAlaLysC-1O-1(+1)",
    ],
    "sequence": "AAK",
    "unsaturation": 2,
  },
  Object {
    "atoms": Object {
      "C": 11,
      "H": 23,
      "N": 4,
      "O": 2,
    },
    "charge": 1,
    "comment": "a3",
    "em": 243.18210099815,
    "ionization": Object {
      "charge": 1,
      "em": 22.989769282,
      "mf": "Na+",
    },
    "mf": "C11H23N4O2(+1)",
    "ms": Object {
      "charge": 2,
      "delta": 16.914613439834056,
      "em": 133.08538656016594,
      "ionization": "Na+",
      "ppm": 127095.94852615324,
    },
    "mw": 243.32635490179035,
    "parts": Array [
      "HAlaAlaLysC-1O-1(+1)",
    ],
    "sequence": "AAK",
    "unsaturation": 2,
  },
]
`;

exports[`fromPeptidicSequence AAKKKKKKKKKKKKKKKKKK allowNeutralLoss 1`] = `
Array [
  Object {
    "atoms": Object {
      "C": 11,
      "H": 20,
      "N": 3,
      "O": 2,
    },
    "charge": 1,
    "comment": "a3",
    "em": 226.15555189703,
    "ionization": Object {
      "charge": 1,
      "em": 1.00782503223,
      "mf": "H+",
    },
    "mf": "C11H20N3O2(+1)",
    "ms": Object {
      "charge": 2,
      "delta": 36.418860115279074,
      "em": 113.58113988472093,
      "ionization": "H+",
      "ppm": 320641.7909896164,
    },
    "mw": 226.29582942817723,
    "parts": Array [
      "HAlaAlaLys(N-1H-3)C-1O-1(+1)",
    ],
    "sequence": "AAK",
    "unsaturation": 3,
  },
  Object {
    "atoms": Object {
      "C": 11,
      "H": 23,
      "N": 4,
      "O": 2,
    },
    "charge": 1,
    "comment": "a3",
    "em": 243.18210099815,
    "ionization": Object {
      "charge": 1,
      "em": 1.00782503223,
      "mf": "H+",
    },
    "mf": "C11H23N4O2(+1)",
    "ms": Object {
      "charge": 2,
      "delta": 27.905585564719075,
      "em": 122.09441443528092,
      "ionization": "H+",
      "ppm": 228557.42986925173,
    },
    "mw": 243.32635490179035,
    "parts": Array [
      "HAlaAlaLysC-1O-1(+1)",
    ],
    "sequence": "AAK",
    "unsaturation": 2,
  },
  Object {
    "atoms": Object {
      "C": 11,
      "H": 20,
      "N": 3,
      "O": 2,
    },
    "charge": 1,
    "comment": "a3",
    "em": 226.15555189703,
    "ionization": Object {
      "charge": 1,
      "em": 22.989769282,
      "mf": "Na+",
    },
    "mf": "C11H20N3O2(+1)",
    "ms": Object {
      "charge": 2,
      "delta": 25.42788799039407,
      "em": 124.57211200960593,
      "ionization": "Na+",
      "ppm": 204121.83417451644,
    },
    "mw": 226.29582942817723,
    "parts": Array [
      "HAlaAlaLys(N-1H-3)C-1O-1(+1)",
    ],
    "sequence": "AAK",
    "unsaturation": 3,
  },
  Object {
    "atoms": Object {
      "C": 11,
      "H": 23,
      "N": 4,
      "O": 2,
    },
    "charge": 1,
    "comment": "a3",
    "em": 243.18210099815,
    "ionization": Object {
      "charge": 1,
      "em": 22.989769282,
      "mf": "Na+",
    },
    "mf": "C11H23N4O2(+1)",
    "ms": Object {
      "charge": 2,
      "delta": 16.914613439834056,
      "em": 133.08538656016594,
      "ionization": "Na+",
      "ppm": 127095.94852615324,
    },
    "mw": 243.32635490179035,
    "parts": Array [
      "HAlaAlaLysC-1O-1(+1)",
    ],
    "sequence": "AAK",
    "unsaturation": 2,
  },
  Object {
    "atoms": Object {
      "C": 17,
      "H": 29,
      "N": 4,
      "O": 3,
    },
    "charge": 1,
    "comment": "a4",
    "em": 337.2239658111,
    "ionization": Object {
      "charge": 1,
      "em": 1.00782503223,
      "mf": "H+",
    },
    "mf": "C17H29N4O3(+1)",
    "ms": Object {
      "charge": 2,
      "delta": -19.11534684175595,
      "em": 169.11534684175595,
      "ionization": "H+",
      "ppm": -113031.4143496539,
    },
    "mw": 337.4378197308548,
    "parts": Array [
      "HAlaAlaLys(N-1H-3)Lys(N-1H-3)C-1O-1(+1)",
    ],
    "sequence": "AAKK",
    "unsaturation": 5,
  },
  Object {
    "atoms": Object {
      "C": 17,
      "H": 32,
      "N": 5,
      "O": 3,
    },
    "charge": 1,
    "comment": "a4",
    "em": 354.25051491221996,
    "ionization": Object {
      "charge": 1,
      "em": 1.00782503223,
      "mf": "H+",
    },
    "mf": "C17H32N5O3(+1)",
    "ms": Object {
      "charge": 2,
      "delta": -27.62862139231592,
      "em": 177.62862139231592,
      "ionization": "H+",
      "ppm": -155541.49537249695,
    },
    "mw": 354.46834520446794,
    "parts": Array [
      "HAlaAlaLysLys(N-1H-3)C-1O-1(+1)",
    ],
    "sequence": "AAKK",
    "unsaturation": 4,
  },
  Object {
    "atoms": Object {
      "C": 17,
      "H": 29,
      "N": 4,
      "O": 3,
    },
    "charge": 1,
    "comment": "a4",
    "em": 337.2239658111,
    "ionization": Object {
      "charge": 1,
      "em": 22.989769282,
      "mf": "Na+",
    },
    "mf": "C17H29N4O3(+1)",
    "ms": Object {
      "charge": 2,
      "delta": -30.10631896664094,
      "em": 180.10631896664094,
      "ionization": "Na+",
      "ppm": -167158.59354283506,
    },
    "mw": 337.4378197308548,
    "parts": Array [
      "HAlaAlaLys(N-1H-3)Lys(N-1H-3)C-1O-1(+1)",
    ],
    "sequence": "AAKK",
    "unsaturation": 5,
  },
  Object {
    "atoms": Object {
      "C": 17,
      "H": 35,
      "N": 6,
      "O": 3,
    },
    "charge": 1,
    "comment": "a4",
    "em": 371.27706401334,
    "ionization": Object {
      "charge": 1,
      "em": 1.00782503223,
      "mf": "H+",
    },
    "mf": "C17H35N6O3(+1)",
    "ms": Object {
      "charge": 2,
      "delta": -36.14189594287595,
      "em": 186.14189594287595,
      "ionization": "H+",
      "ppm": -194163.14505557274,
    },
    "mw": 371.49887067808106,
    "parts": Array [
      "HAlaAlaLysLysC-1O-1(+1)",
    ],
    "sequence": "AAKK",
    "unsaturation": 3,
  },
  Object {
    "atoms": Object {
      "C": 17,
      "H": 32,
      "N": 5,
      "O": 3,
    },
    "charge": 1,
    "comment": "a4",
    "em": 354.25051491221996,
    "ionization": Object {
      "charge": 1,
      "em": 22.989769282,
      "mf": "Na+",
    },
    "mf": "C17H32N5O3(+1)",
    "ms": Object {
      "charge": 2,
      "delta": -38.61959351720091,
      "em": 188.6195935172009,
      "ionization": "Na+",
      "ppm": -204748.5777965005,
    },
    "mw": 354.46834520446794,
    "parts": Array [
      "HAlaAlaLysLys(N-1H-3)C-1O-1(+1)",
    ],
    "sequence": "AAKK",
    "unsaturation": 4,
  },
  Object {
    "atoms": Object {
      "C": 17,
      "H": 35,
      "N": 6,
      "O": 3,
    },
    "charge": 1,
    "comment": "a4",
    "em": 371.27706401334,
    "ionization": Object {
      "charge": 1,
      "em": 22.989769282,
      "mf": "Na+",
    },
    "mf": "C17H35N6O3(+1)",
    "ms": Object {
      "charge": 2,
      "delta": -47.13286806776094,
      "em": 197.13286806776094,
      "ionization": "Na+",
      "ppm": -239091.8801605415,
    },
    "mw": 371.49887067808106,
    "parts": Array [
      "HAlaAlaLysLysC-1O-1(+1)",
    ],
    "sequence": "AAKK",
    "unsaturation": 3,
  },
  Object {
    "atoms": Object {
      "C": 23,
      "H": 38,
      "N": 5,
      "O": 4,
    },
    "charge": 1,
    "comment": "a5",
    "em": 448.29237972516995,
    "ionization": Object {
      "charge": 1,
      "em": 1.00782503223,
      "mf": "H+",
    },
    "mf": "C23H38N5O4(+1)",
    "ms": Object {
      "charge": 2,
      "delta": -74.64955379879092,
      "em": 224.64955379879092,
      "ionization": "H+",
      "ppm": -332293.3544112505,
    },
    "mw": 448.57981003353234,
    "parts": Array [
      "HAlaAlaLys(N-1H-3)Lys(N-1H-3)Lys(N-1H-3)C-1O-1(+1)",
    ],
    "sequence": "AAKKK",
    "unsaturation": 7,
  },
  Object {
    "atoms": Object {
      "C": 23,
      "H": 41,
      "N": 6,
      "O": 4,
    },
    "charge": 1,
    "comment": "a5",
    "em": 465.31892882629,
    "ionization": Object {
      "charge": 1,
      "em": 1.00782503223,
      "mf": "H+",
    },
    "mf": "C23H41N6O4(+1)",
    "ms": Object {
      "charge": 2,
      "delta": -83.16282834935095,
      "em": 233.16282834935095,
      "ionization": "H+",
      "ppm": -356672.75499312003,
    },
    "mw": 465.61033550714546,
    "parts": Array [
      "HAlaAlaLysLys(N-1H-3)Lys(N-1H-3)C-1O-1(+1)",
    ],
    "sequence": "AAKKK",
    "unsaturation": 6,
  },
  Object {
    "atoms": Object {
      "C": 23,
      "H": 38,
      "N": 5,
      "O": 4,
    },
    "charge": 1,
    "comment": "a5",
    "em": 448.29237972516995,
    "ionization": Object {
      "charge": 1,
      "em": 22.989769282,
      "mf": "Na+",
    },
    "mf": "C23H38N5O4(+1)",
    "ms": Object {
      "charge": 2,
      "delta": -85.64052592367591,
      "em": 235.6405259236759,
      "ionization": "Na+",
      "ppm": -363437.1701895408,
    },
    "mw": 448.57981003353234,
    "parts": Array [
      "HAlaAlaLys(N-1H-3)Lys(N-1H-3)Lys(N-1H-3)C-1O-1(+1)",
    ],
    "sequence": "AAKKK",
    "unsaturation": 7,
  },
  Object {
    "atoms": Object {
      "C": 23,
      "H": 44,
      "N": 7,
      "O": 4,
    },
    "charge": 1,
    "comment": "a5",
    "em": 482.34547792740995,
    "ionization": Object {
      "charge": 1,
      "em": 1.00782503223,
      "mf": "H+",
    },
    "mf": "C23H44N7O4(+1)",
    "ms": Object {
      "charge": 2,
      "delta": -91.67610289991092,
      "em": 241.67610289991092,
      "ionization": "H+",
      "ppm": -379334.5796289928,
    },
    "mw": 482.6408609807586,
    "parts": Array [
      "HAlaAlaLysLysLys(N-1H-3)C-1O-1(+1)",
    ],
    "sequence": "AAKKK",
    "unsaturation": 5,
  },
  Object {
    "atoms": Object {
      "C": 23,
      "H": 41,
      "N": 6,
      "O": 4,
    },
    "charge": 1,
    "comment": "a5",
    "em": 465.31892882629,
    "ionization": Object {
      "charge": 1,
      "em": 22.989769282,
      "mf": "Na+",
    },
    "mf": "C23H41N6O4(+1)",
    "ms": Object {
      "charge": 2,
      "delta": -94.15380047423594,
      "em": 244.15380047423594,
      "ionization": "Na+",
      "ppm": -385633.1553772861,
    },
    "mw": 465.61033550714546,
    "parts": Array [
      "HAlaAlaLysLys(N-1H-3)Lys(N-1H-3)C-1O-1(+1)",
    ],
    "sequence": "AAKKK",
    "unsaturation": 6,
  },
  Object {
    "atoms": Object {
      "C": 23,
      "H": 47,
      "N": 8,
      "O": 4,
    },
    "charge": 1,
    "comment": "a5",
    "em": 499.37202702852994,
    "ionization": Object {
      "charge": 1,
      "em": 1.00782503223,
      "mf": "H+",
    },
    "mf": "C23H47N8O4(+1)",
    "ms": Object {
      "charge": 2,
      "delta": -100.18937745047091,
      "em": 250.18937745047091,
      "ionization": "H+",
      "ppm": -400454.1618490779,
    },
    "mw": 499.6713864543717,
    "parts": Array [
      "HAlaAlaLysLysLysC-1O-1(+1)",
    ],
    "sequence": "AAKKK",
    "unsaturation": 4,
  },
  Object {
    "atoms": Object {
      "C": 23,
      "H": 44,
      "N": 7,
      "O": 4,
    },
    "charge": 1,
    "comment": "a5",
    "em": 482.34547792740995,
    "ionization": Object {
      "charge": 1,
      "em": 22.989769282,
      "mf": "Na+",
    },
    "mf": "C23H44N7O4(+1)",
    "ms": Object {
      "charge": 2,
      "delta": -102.6670750247959,
      "em": 252.6670750247959,
      "ionization": "Na+",
      "ppm": -406333.41330571263,
    },
    "mw": 482.6408609807586,
    "parts": Array [
      "HAlaAlaLysLysLys(N-1H-3)C-1O-1(+1)",
    ],
    "sequence": "AAKKK",
    "unsaturation": 5,
  },
  Object {
    "atoms": Object {
      "C": 23,
      "H": 47,
      "N": 8,
      "O": 4,
    },
    "charge": 1,
    "comment": "a5",
    "em": 499.37202702852994,
    "ionization": Object {
      "charge": 1,
      "em": 22.989769282,
      "mf": "Na+",
    },
    "mf": "C23H47N8O4(+1)",
    "ms": Object {
      "charge": 2,
      "delta": -111.18034957535588,
      "em": 261.1803495753559,
      "ionization": "Na+",
      "ppm": -425684.20540105784,
    },
    "mw": 499.6713864543717,
    "parts": Array [
      "HAlaAlaLysLysLysC-1O-1(+1)",
    ],
    "sequence": "AAKKK",
    "unsaturation": 4,
  },
  Object {
    "atoms": Object {
      "C": 29,
      "H": 47,
      "N": 6,
      "O": 5,
    },
    "charge": 1,
    "comment": "a6",
    "em": 559.36079363924,
    "ionization": Object {
      "charge": 1,
      "em": 1.00782503223,
      "mf": "H+",
    },
    "mf": "C29H47N6O5(+1)",
    "ms": Object {
      "charge": 2,
      "delta": -130.18376075582592,
      "em": 280.1837607558259,
      "ionization": "H+",
      "ppm": -464637.06677589443,
    },
    "mw": 559.72180033621,
    "parts": Array [
      "HAlaAlaLys(N-1H-3)Lys(N-1H-3)Lys(N-1H-3)Lys(N-1H-3)C-1O-1(+1)",
    ],
    "sequence": "AAKKKK",
    "unsaturation": 9,
  },
  Object {
    "atoms": Object {
      "C": 29,
      "H": 50,
      "N": 7,
      "O": 5,
    },
    "charge": 1,
    "comment": "a6",
    "em": 576.38734274036,
    "ionization": Object {
      "charge": 1,
      "em": 1.00782503223,
      "mf": "H+",
    },
    "mf": "C29H50N7O5(+1)",
    "ms": Object {
      "charge": 2,
      "delta": -138.69703530638594,
      "em": 288.69703530638594,
      "ionization": "H+",
      "ppm": -480424.17601964885,
    },
    "mw": 576.7523258098231,
    "parts": Array [
      "HAlaAlaLysLys(N-1H-3)Lys(N-1H-3)Lys(N-1H-3)C-1O-1(+1)",
    ],
    "sequence": "AAKKKK",
    "unsaturation": 8,
  },
  Object {
    "atoms": Object {
      "C": 29,
      "H": 47,
      "N": 6,
      "O": 5,
    },
    "charge": 1,
    "comment": "a6",
    "em": 559.36079363924,
    "ionization": Object {
      "charge": 1,
      "em": 22.989769282,
      "mf": "Na+",
    },
    "mf": "C29H47N6O5(+1)",
    "ms": Object {
      "charge": 2,
      "delta": -141.1747328807109,
      "em": 291.1747328807109,
      "ionization": "Na+",
      "ppm": -484845.4104654324,
    },
    "mw": 559.72180033621,
    "parts": Array [
      "HAlaAlaLys(N-1H-3)Lys(N-1H-3)Lys(N-1H-3)Lys(N-1H-3)C-1O-1(+1)",
    ],
    "sequence": "AAKKKK",
    "unsaturation": 9,
  },
  Object {
    "atoms": Object {
      "C": 29,
      "H": 53,
      "N": 8,
      "O": 5,
    },
    "charge": 1,
    "comment": "a6",
    "em": 593.41389184148,
    "ionization": Object {
      "charge": 1,
      "em": 1.00782503223,
      "mf": "H+",
    },
    "mf": "C29H53N8O5(+1)",
    "ms": Object {
      "charge": 2,
      "delta": -147.2103098569459,
      "em": 297.2103098569459,
      "ionization": "H+",
      "ppm": -495306.87521506764,
    },
    "mw": 593.7828512834362,
    "parts": Array [
      "HAlaAlaLysLysLys(N-1H-3)Lys(N-1H-3)C-1O-1(+1)",
    ],
    "sequence": "AAKKKK",
    "unsaturation": 7,
  },
  Object {
    "atoms": Object {
      "C": 29,
      "H": 50,
      "N": 7,
      "O": 5,
    },
    "charge": 1,
    "comment": "a6",
    "em": 576.38734274036,
    "ionization": Object {
      "charge": 1,
      "em": 22.989769282,
      "mf": "Na+",
    },
    "mf": "C29H50N7O5(+1)",
    "ms": Object {
      "charge": 2,
      "delta": -149.68800743127093,
      "em": 299.68800743127093,
      "ionization": "Na+",
      "ppm": -499479.4710482357,
    },
    "mw": 576.7523258098231,
    "parts": Array [
      "HAlaAlaLysLys(N-1H-3)Lys(N-1H-3)Lys(N-1H-3)C-1O-1(+1)",
    ],
    "sequence": "AAKKKK",
    "unsaturation": 8,
  },
]
`;

exports[`fromPeptidicSequence Linked AA(H-1#1)AA,GG(H-1#1)GG 1`] = `
Array [
  Array [
    "#1C6H4#2",
    "HAlaAla(H-1#1)AlaAlaOH",
    "HGlyGly(H-1#2)GlyGlyOH",
  ],
  Array [
    "#1C6H4#2",
    "HAlaAla(H-1#1)C-1O-1(+1)",
    "HGlyGly(H-1#2)C-1O-1(+1)",
  ],
  Array [
    "#1C6H4#2",
    "HAlaAla(H-1#1)C-1O-1(+1)",
    "HGlyGly(H-1#2)GlyC-1O-1(+1)",
  ],
  Array [
    "#1C6H4#2",
    "HAlaAla(H-1#1)AlaC-1O-1(+1)",
    "HGlyGly(H-1#2)C-1O-1(+1)",
  ],
  Array [
    "#1C6H4#2",
    "HAlaAla(H-1#1)AlaC-1O-1(+1)",
    "HGlyGly(H-1#2)GlyC-1O-1(+1)",
  ],
  Array [
    "#1C6H4#2",
    "HAlaAla(H-1#1)C-1O-1(+1)",
    "HGlyGly(H-1#2)GlyGlyOH",
  ],
  Array [
    "#1C6H4#2",
    "HAlaAla(H-1#1)AlaAlaOH",
    "HGlyGly(H-1#2)C-1O-1(+1)",
  ],
  Array [
    "#1C6H4#2",
    "HAlaAla(H-1#1)AlaC-1O-1(+1)",
    "HGlyGly(H-1#2)GlyGlyOH",
  ],
  Array [
    "#1C6H4#2",
    "HAlaAla(H-1#1)AlaAlaOH",
    "HGlyGly(H-1#2)GlyC-1O-1(+1)",
  ],
]
`;

exports[`fromPeptidicSequence Optional linked AA(H-1#1)A,GG(H-1#1)G 1`] = `
Array [
  Array [
    "#1C6H4#2",
    "HAlaAla(H-1#1)AlaOH",
    "HGlyGly(H-1#2)GlyOH",
  ],
  Array [
    ,
    "HGlyC-1O-1(+1)",
  ],
  Array [
    ,
    "HAlaC-1O-1(+1)",
  ],
  Array [
    "#1C6H4#2",
    "HAlaAla(H-1#1)C-1O-1(+1)",
    "HGlyGly(H-1#2)C-1O-1(+1)",
  ],
  Array [
    "#1C6H4#2",
    "HAlaAla(H-1#1)C-1O-1(+1)",
    "HGlyGly(H-1#2)GlyOH",
  ],
  Array [
    "#1C6H4#2",
    "HAlaAla(H-1#1)AlaOH",
    "HGlyGly(H-1#2)C-1O-1(+1)",
  ],
]
`;
